python/python/lance/arrow.py

```diff
@@ -11,14 +11,182 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
-from typing import Optional, Union
+from typing import List, Optional, Union
 
 import pyarrow as pa
 
 from .lance import BFloat16
 from .lance import bfloat16_array as bfloat16_array
 
 
+class ImageURIType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.string(), "lance.arrow.image_uri")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return ImageURIType()
+
+    def __arrow_ext_class__(self):
+        return ImageURIArray
+
+    def __arrow_ext_scalar_class__(self):
+        return ImageURIScalar
+
+
+class EncodedImageType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.binary(), "lance.arrow.encoded_image")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return EncodedImageType()
+
+    def __arrow_ext_class__(self):
+        return EncodedImageArray
+
+    def __arrow_ext_scalar_class__(self):
+        return EncodedImageScalar
+
+
+class ImageTensorType(pa.ExtensionType):
+    def __init__(self, arrow_type, shape):
+        # TODO: use pa.VariableShapeTensorArray once available?
+        pa.ExtensionType.__init__(
+            self, pa.fixed_shape_tensor(arrow_type, shape), "lance.arrow.image_tensor"
+        )
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return ImageTensorType()
+
+    def __arrow_ext_class__(self):
+        return ImageTensorArray
+
+    def __arrow_ext_scalar_class__(self):
+        return ImageTensorScalar
+
+
+class ImageURIArray(pa.ExtensionArray):
+    def __repr__(self):
+        return "<lance.arrow.ImageURIArray object at 0x%016x>\n%s" % (
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    @classmethod
+    def from_uris(cls, uris: Union[pa.StringArray, List[str]]):
+        if isinstance(uris, list):
+            return cls.from_storage(ImageURIType(), pa.array(uris, type=pa.string()))
+        else:
+            return cls.from_storage(ImageURIType(), uris)
```

Comment:
We are passing anything that isn't a list into `cls.from_storage()`, which might not be super safe. The typehint won't guarantee someone will pass something weird.

Also, maybe we could take an arbitrary iterable?

```suggestion
    def from_uris(cls, uris: Union[pa.StringArray, Iterable[Union[str, Path]]]):
        if isinstance(uris, pa.StringArray):
            return cls.from_storage(ImageURIType(), uris)
        elif isinstance(uris, Iterable):
            storage = pa.array((str(uri) for uri in uris), type=pa.string())
            return cls.from_storage(ImageURIType(), storage)
        else:
            raise TypeError("Cannot build a ImageURIArray from {}".format(type(uris))
```

Reply:
Shall we add a check that the string is a valid URI? I'm not sure how deep we want to go here.

Reply:
I don't think we have to.